Content/posts/jwtkit-v5.md

```diff
@@ -0,0 +1,182 @@
+
+```diff
+- let payload = signers.verify(token, as: YourPayload.self)
++ let payload = try await keyCollection.verify(token, as: YourPayload.self)
+```
+
+### JWT
+
+If you want to try out version 5 using the Vapor integration, you need to update your dependency to:
+
+```swift
+.package(url: "https://github.com/vapor/jwt", from: "5.0.0-beta.1"),
+```
+
+and then migrate to the new API:
+
+```diff
+- app.jwt.signers.use(.hs256(key: "secret"), kid: "foo")
++ await app.jwt.keys.addHS256(key: "secret", kid: "foo")
+```
+
+After adding a key, you can create your payload like:
+
+```diff
+struct SomePayload: JWTPayload {
+    var exampleName: String
+    
+    // the claims did not change
+
+-   func verify(using signer: JWTSigner) throws {
++   func verify(using signer: JWTAlgorithm) async throws {
+        // ... 
+    }
+}
+```
+
+Adn then you can sign and verify your tokens like this:
+
+```diff
+- let token = try signers.sign(SomePayload(exampleName: "bob"), kid: "foo")
++ let token = try await app.jwt.keys.sign(SomePayload(exampleName: "bob"), header: ["kid": "foo"])
+
+- let payload = try signers.verify(token, as: SomePayload.self)
++ let payload = try await app.jwt.keys.verify(token, as: SomePayload.self)
+```
+
+The rest of the methods are basically the same, but asynchronous.
+
+## Customization
+
+Then, we've added some cool customization features which weren't available before. Custom headers are now a thing, as the header is now a dictionary which you can fill however you want. To access custom fields you don't even need to use the dictionary syntax as the header is accessible using `@_dynamicMemberLookup`. For more traditional users, the usual fields are type-safely provided as extensions. A cool example of custom header use is the [Open Banking spec](// https://openbanking.atlassian.net/wiki/spaces/DZ/pages/937656404/Read+Write+Data+API+Specification+-+v3.1):
+
+```swift
+let customFields: JWTHeader = [
+    "kid": "90210ABAD",
+    "http://openbanking.org.uk/iat": 1_501_497_671,
+    "http://openbanking.org.uk/iss": "C=UK, ST=England, L=London, O=Acme Ltd.",
+    "http://openbanking.org.uk/tan": "openbanking.org.uk",
+    "crit": [
+        "b64",
+        "http://openbanking.org.uk/iat",
+        "http://openbanking.org.uk/iss",
+        "http://openbanking.org.uk/tan"
+    ],
+]
+let token = try await keyCollection.sign(payload, header: customFields)
+```
+
+Parsing and serializing are now also customizable, meaning that you can define your own implementation for parsers and serializers using the custom headers you defined, for example compressing them with `zip` or `deflate` or using a non-encoded payload by setting the `b64` header to false. Rather than implementing the whole JOSE standard, we decided to let users extend the package however they want to. In the tests there's an example which shows how to set the `b64` header to `false` (which by default is non present, meaning `true`):
```

Comment:
```suggestion
Parsing and serializing are now also customizable, meaning that you can define your own implementation for parsers and serializers using the custom headers you defined, for example compressing them with `zip` or `deflate`, or using an unencoded payload by setting the `b64` header to `false`. Rather than implementing the entire JOSE standard, we decided to let users extend the package as desired. The tests include an example which shows how to set the `b64` header (which by default is not present and thus assumed to have the value `true`):
```